qiskit_experiments/data_processing/nodes.py

```diff
@@ -0,0 +1,173 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Different data analysis steps."""
+
+from abc import abstractmethod
+from typing import Any, Dict, Optional, Tuple
+import numpy as np
+
+from qiskit_experiments.data_processing.data_action import DataAction
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class IQPart(DataAction):
+    """Abstract class for IQ data post-processing."""
+
+    def __init__(self, scale: Optional[float] = None, validate: bool = True):
+        """
+        Args:
+            scale: Float with which to multiply the IQ data.
+            validate: If set to False the DataAction will not validate its input.
+        """
+        self.scale = scale
+        super().__init__(validate)
+
+    @abstractmethod
+    def _process(self, datum: np.array) -> np.array:
+        """Defines how the IQ point will be processed.
+
+        Args:
+            datum: A 3D array of shots, qubits, and a complex IQ point as [real, imaginary].
+
+        Returns:
+            Processed IQ point.
+        """
+
+    def _format_data(self, datum: Any, validate: bool = True) -> Any:
+        """Check that the IQ data has the correct format and convert to numpy array.
+
+        Args:
+            datum: A single item of data which corresponds to single-shot IQ data. It should
+                have dimension three: shots, qubits, iq-point as [real, imaginary].
+            validate: If True the DataAction checks that the format of the datum is valid.
+
+        Returns:
+            datum as a numpy array.
+
+        Raises:
+            DataProcessorError: If the datum does not have the correct format.
+        """
+        datum = np.asarray(datum, dtype=complex)
```

Comment:
Why do you cast to complex if you are taking real valued I and Q parts?

Reply:
I think we can stick to float here.

Reply:
done here 657f17b